### PR TITLE
Feature/vertical merge

### PR DESCRIPTION
**Pull request by kaniak274**

### CODE REVIEW

File: static/merge.js [hunk 1]
```diff
@@ -51,7 +57,7 @@
         const allHeaders = $(`.star_chart_table-${constructId}`)
             .find('th');
     
-        return startIndex + (headers.length) === allHeaders.length - cells.length - levelCount - 1
+        return startIndex + (headers.length) === allHeaders.length - (2 * cells.length) - levelCount - 2
```
Comment (rkowalski4) [inline static/merge.js]: That's quite complicated. Could you elaborate?

File: static/merge.js [hunk 2]
```diff
@@ -104,11 +110,11 @@
         const cellIndex = headers[i].cellIndex;
 
         if (i == 0) {
-            startIndex = cellIndex;
+            startIndex = cellIndex - 1;
         }
                 
         const cells = $(`.star_chart_table-${constructId} .${construct}`)
-            .find(`td:eq(${cellIndex - 1})`);
+            .find(`td:eq(${cellIndex - 2})`);
```
Comment (rkowalski4) [inline static/merge.js]: 2? Why?

File: static/merge.js [hunk 3]
```diff
@@ -400,4 +442,162 @@
...
+
+        var size = mergedVertical[constructId][classId].length;
+
+        for (var i = 0; i < size; i++) {
+            for (var j = 0; j < sublevels; j++) {
+                const observations = $(mergedVertical[constructId][classId][i].cells[j]).data('modal-launch-observations');
+                
+                if (observations) {
+                    // Add observations and filter reoccurring ones.
+                    observations.map(observation => {
```
Comment (rkowalski4) [inline static/merge.js]: You call map for side effects. You should use `forEach` insted.

File: static/merge.js [hunk 4]
```diff
@@ -400,4 +442,162 @@
...
+            var cells = $($(students[i]).parent()).find('td')
+
+            sublevels = cells.length;
+
+            let data = {
+                'student': students[i],
+                'cells': cells
+            };
+
+            mergedVertical[constructId][classId].push(data);
```
Comment (rkowalski4) [inline static/merge.js]: You can create dict in push method to save a few lines - `data` name has no information in itself

File: static/merge.js [hunk 5]
```diff
@@ -400,4 +442,162 @@
...
+        cells.remove();
+
+        // Remove header with class name
+        $(tr).find('th:eq(1)').remove();
+
+        // Append to row with button header with first student <th> element.
+        $(tr).append(mergedVertical[constructId][classId][size - 1].student);
+
+        // Append saved <td>s elements to row.
+        [...Array(sublevels).keys()].map(
```
Comment (rkowalski4) [inline static/merge.js]: foreach instead of map

File: static/merge.js [hunk 6]
```diff
@@ -400,4 +442,162 @@
...
+            i => $(tr).append(mergedVertical[constructId][classId][size - 1].cells[i]));
```
Comment (rkowalski4) [inline static/merge.js]: is `i` a number? If not, please rename it to some more verbose name such as `key`. Take a note you have similar code below, so apply this change there as well.

File: templates/_star_chart_v1.html [hunk 7]
```diff
@@ -1,67 +1,113 @@
...
+                            </th>
+                    
+                            {% for sublevel, observations in sublevels.items %}
+                                <td data-csl-id="{{ csl_id }}"
+                                    class="text-center"
+                                    {% if observations|length > 0 %}
+                                        data-modal-launch-observations="{{ observations|observation_pks }}"
+                                    {% endif %}>
+
+                                    {% if observations|length > 0 %}
```
Comment (rkowalski4) [inline templates/_star_chart_v1.html]: I think that empty collection of observation returns false, so you don't have to calculate the length.

File: static/merge.js (hunk 6 above)
Comment (kaniak274) [reply static/merge.js]: Yeah i is a number. `[...Array(sublevels).keys()]` this creates for me array with range from 0 to sublevels value.

**Pull request merged**